### ISSUE TITLE
Error in Modelica.Fluid.Utilities.regFun3

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#495]
**Modified by dietmarw on 5 Dec 2011 12:51 UTC**
There seems to be an error in the Modelica.Fluid function Modelica.Fluid.Utilities.regFun3
```text/x-modelica
model TestRegFun3
Real x;
//parameter Real a = 15;
//parameter Real b = 1.0;

Real a=15;
Real b=1.0;

equation
x = Modelica.Fluid.Utilities.regFun3(
            time,
            0,
            100,
            50,
            800/a,
            0,
            b/a);
end TestRegFun3;
```
Calling regFun3 with:
```text/x-modelica
parameter a = 15;
parameter b = 1.0;
```
gives a different result than calling it with:
```text/x-modelica
Real a=15;
Real b=1.0;
```
According to the Dymola developers this is a problem in the function rather than in Dymola (see the email from Hans Olsson below):

The reason for the difference between parameters and non-parameters is that non-
parameters are evaluated and this causes slight round-off differences that
keeps the code out of the bad code-sections.

The bad sections are trigged by a=15, b=1 and triggers:
```
if abs(y1d+y0d-2*Delta0)<100*Modelica.Constants.eps then
  ...
elseif abs(y1d+y0d-2*Delta0)<100*Modelica.Constants.eps then
```
Both of these simulate "infinite slopes" by using Modelica.Constants.Inf.
Changing from Inf to just 1e9 gives somewhat correct result with a large noise -
indicating that the actual problem is some form of truncation.

I believe the error is found there and that you can contact the library authors.

----

**Reported by erik&#x2e;aberg on 4 Feb 2011 13:40 UTC**
There seems to be an error in the Modelica.Fluid function Modelica.Fluid.Utilities.regFun3

model TestRegFun3
Real x;
//parameter Real a = 15;
//parameter Real b = 1.0;

Real a=15;
Real b=1.0;

equation
x = Modelica.Fluid.Utilities.regFun3(
            time,
            0,
            100,
            50,
            800/a,
            0,
            b/a);
end TestRegFun3;

Calling regFun3 with:

parameter a = 15;
parameter b = 1.0;

gives a different result than calling it with:

Real a=15;
Real b=1.0;

According to the Dymola developers this is a problem in the function rather than in Dymola (see the email from Hans Olsson below):

The reason for the difference between parameters and non-parameters is that non-
parameters are evaluated and this causes slight round-off differences that
keeps the code out of the bad code-sections.

The bad sections are trigged by a=15, b=1 and triggers:

if abs(y1d+y0d-2\*Delta0)<100\*Modelica.Constants.eps then
  ...

elseif abs(y1d+y0d-2\*Delta0)<100\*Modelica.Constants.eps then

Both of these simulate "infinite slopes" by using Modelica.Constants.Inf.
Changing from Inf to just 1e9 gives somewhat correct result with a large noise -
indicating that the actual problem is some form of truncation.

I believe the error is found there and that you can contact the library authors.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/495

**Comment by modelica-trac-importer** [comment]
**Modified by hansolsson on 17 Nov 2011 10:48 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 5 Dec 2011 12:20 UTC**
To clarify a bit:

There is at least one numerically not well-designed part of the code in this function.

That part of the code is an attempt to guard against values that are too close together for normal floating point numbers (=IEEE double precision arithmetic). Unfortunately the attempt does not work due to truncation for the floating point numbers (at least I think truncation is the issue).

The issue is normally avoided in Dymola due to specifics in the code-generation, i.e. slight rounding of values. 
That rounding has no impact on the bad code itself, it just avoids running it.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 5 Dec 2011 12:51 UTC**
Unhiding so reporter has access to his own ticket.

**Comment by modelica-trac-importer** [comment]
**Modified by hansolsson on 13 Jan 2012 15:36 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 14 Jan 2012 23:08 UTC**
I propose to get this fixed for the next 3.2.1 release (and also for the 3.2 maintenance).

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 24 Jan 2012 09:21 UTC**
Two observations:

Using 10<sup>6</sup> or 10<sup>3</sup> instead of Modelica.Constants.inf gives fairly good result.
There are two uses of Modelica.Constants.inf, they must be replaced by the same (smaller) number or the results will be incorrect.

**Comment by modelica-trac-importer** [comment]
**Comment by msielemann on 9 Feb 2012 11:56 UTC**
The issue was fixed in 88f07f19d5f27c492cc93601f8500b6e2888f20e (trunk) and e90944a43b079f35dab26ad6ccfa14b78b64cbd7 (3.2 maintenance).

The issue was due to division by zero in the computation of xstar, the inflection point of the cubic polynomial. This variable is used in the co-monotonicity criteria of _[Gasparo and Morandi 1991]_; therefore, the criteria did not evaluate properly.

The problem was solved by arguing that if the inflection point is at +/- infinity, i.e., outside of the interpolation interval, that then the cubic S0 is co-monotone if the signs of y0d and y1d are identical (also allowing that they are zero). Therefore, the cubic S0 can be returned directly.

In the mentioned commits, a few minor changes were also made.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 9 Feb 2012 17:52 UTC**
Reopened since in both commits illegal html code was added. Please remove the `<p>...</p>` environments around the `<ul>...</ul>` environments (at least in trunk) as that is illegal html. Thanks.

**Comment by modelica-trac-importer** [comment]
**Comment by msielemann on 10 Feb 2012 09:25 UTC**
Thanks for pointing this out. Corrected in 3f7c4485f30933175f48ad609d573c6238eb74a8.
